File: NeetCode/Graph/Course_Schedule/course__schedule.py
from typing import Optional, List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        courses = [[] for _ in range(numCourses)]

        for a, b in prerequisites:
            courses[a].append(b)

        checking = set()
        seen = set()

        def dfs(course):
            if course in checking:
                return False
            if course in seen:
                return True
            checking.add(course)

            for pre in courses[course]:
                check = dfs(pre)
                if not check:
                    return False

            checking.remove(course)
            seen.add(course)
            return True

        for c in range(numCourses):
            if not dfs(c):
                return False
            else:
                courses[c] = []
        return True


if __name__ == "__main__":
    sol = Solution()
    numCourses = 5
    prerequisites = [[1, 4], [2, 4], [3, 1], [3, 2]]
    sol.canFinish(numCourses, prerequisites)
    print("Running Solution...")

File: NeetCode/Graph/Course_Schedule/Course_Schedule.cpp
// Course_Schedule.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        vector<vector<int>> courses(numCourses);
        for (vector<int> curr : prerequisites)
        {
            int a = curr[0];
            int b = curr[1];
            courses[a].push_back(b);
        }

        unordered_set<int> checking;

        for (int c = 0; c < numCourses; c++)
        {
            if (!dfs(c, courses, checking, prerequisites))
            {
                return false;
            }
        }
        return true;
    }

    bool dfs(int course, vector<vector<int>> &courses, unordered_set<int> &checking, vector<vector<int>> &prerequisites)
    {
        if (checking.count(course))
        {
            return false;
        }

        checking.insert(course);

        for (int pre : courses[course])
        {
            if (!dfs(pre, courses, checking, prerequisites))
            {
                return false;
            }
        }

        checking.erase(course);
        return true;
    }
};
